Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm

```diff
@@ -1224,6 +1224,61 @@ static _WKWebExtensionContextError toAPI(WebExtensionContext::Error error)
     }).get()];
 }
 
+bool WebExtensionContext::needsPermission(const String& permission, WebExtensionTab* tab, OptionSet<PermissionStateOptions> options)
+{
+    ASSERT(!permission.isEmpty());
+    ASSERT(!options.contains(PermissionStateOptions::SkipRequestedPermissions));
+
+    switch (permissionState(permission, tab, options)) {
```

Comment:
You could refactor the logic in the switch statement here and in the other `needsPermission()` function into shared code, e.g.
```suggestion
    return needsPermission(permissionState(permission, tab, options);
```